plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/scan_completion/trigger.py

```diff
@@ -0,0 +1,283 @@
+        except ValueError as error:
+            raise PluginException(cause="Failed to load cache file", assistance=f"Exception returned was {error}")
+
+
+class Util:
+    @staticmethod
+    def filter_results(params: dict, csv_row: dict):
+        """
+        Filter the outputted results based on the user inputs.
+
+        :param params: Input params
+        :param csv_row: Dict row of the csv results
+
+        :return: New object containing only the necessary fields for the required output.
+        """
+
+        # Input retrieval
+        asset_group = params.get(Input.ASSET_GROUP, None)
+        cve = params.get(Input.CVE, None)
+        hostname = params.get(Input.HOSTNAME, None)
+        source = params.get(Input.SOURCE, None)
+        ip_address = params.get(Input.IP_ADDRESS, None)
+
+        try:
+            new_dct = {
+                "asset_id": int(csv_row["asset_id"]),
+                "ip_address": csv_row["ip_address"],
+                "hostname": csv_row["host_name"],
+                "vulnerability_id": csv_row["vulnerability_id"],
+                "nexpose_id": csv_row["nexpose_id"],
+                "solution_id": Util.strip_msft_id(csv_row.get("solution_id", "")),
+                "solution_summary": csv_row["summary"],
+            }
+        except KeyError as error:
+            raise PluginException(cause="Unable to filter results.", assistance=f"Error: {error}")
+
+        # If an input and it is not found, return None in place of the row to filter
+        # out the result
+        if asset_group and asset_group not in csv_row["asset_group_id"]:
+            return None
+        if cve and cve not in csv_row["nexpose_id"]:
+            return None
+        if hostname and hostname not in csv_row["host_name"]:
+            return None
+        if source and source not in csv_row["source"]:
+            return None
+        if ip_address and ip_address not in csv_row["ip_address"]:
+            return None
+        # Otherwise, return the newly filtered result.
+        else:
+            return new_dct
+
+    @staticmethod
+    def condense_results(results: list):
+        """
+        Helper method to condense vulnerability info into a nested object for the 'vulnerability info' key
+        within the original objects.
+
+        :param results:
+        :return:
+        """
+        merge_keys = ("asset_id", "ip_address", "hostname")
+        dct = {}
+        new_results = []
+
+        for element in results:
+            key = tuple(element[key_] for key_ in merge_keys)
+            partial_element = {key_: value_ for key_, value_ in element.items() if key_ not in merge_keys}
+            dct.setdefault(key, []).append(partial_element)
+
+        for key, value in dct.items():
+            entry = dict(zip(merge_keys, key))
+            entry["vuln_info"] = value
+            new_results.append(entry)
+        return new_results
```

Comment:
I thought about this, but the logic is that there is multiple entries for vuln info per asset.
So originally the data is like this:
```
{asset_id: 1, vuln_id: 1}
{asset_id: 1, vuln_id: 2}
{asset_id: 1, vuln_id: 3}
etc...
```
It then gets condensed into:
```
{asset_id: 1, vuln_info: [{vuln_id: 1, vuln_id: 2, vuln_id: 3}]
```
If I did it how you suggested above, we'd still have multiple entries for the same asset

Reply:
My bad, hadn't realised the database data would be returning the asset ID multiple times. only other suggestion that could maybe be used, but might not fit in the logic of how data is being returned (but it might solve  the issue of having to do the couple of loops in `condense_results`.

In `filter_results` pass a dictionary with the key: value, being `asset_id: dict_values`, and then to return a list at the end we could do `list(results.values())`? eg:

```
results = {}. # dict to be cast to list after filtering
for row in csv_report:
      results = Util.filter_results(results, params, csv_row)
return list(results.values())

def filter_results(results: dict, params: dict, csv_row: dict):
   <filter out rows not needed based on params>
   new_dict = {}. # same as before create new dict
   existing_asset = results.get(asset_id)
   if existing_asset:
       existing_asset['vuln_info'].append(new_dict['vuln_info'])
   else:
       existing_asset = new_dict
   results[asset_id] = existing_asset
```

Reply:
This was awesome ty :)